flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,276 @@
+  private Committer<SinkV2Committable> sinkV2Committer;
+  private final String sinkId;
+  private CommonCommitter commonCommitter;
+
+  // FlinkWriteConf properties stored here for serialization
+  private boolean upsertMode;
+  private FileFormat dataFileFormat;
+  private long targetDataFileSize;
+
+  private IcebergSink(SinkBuilder builder) {
+    super(builder);
+
+    this.tableLoader = builder.tableLoader();
+    this.snapshotProperties = builder.snapshotProperties();
+
+    // We generate a random UUID every time when a sink is created.
+    // This is used to separate files generated by different sinks writing the same table.
+    // Also used to generate the aggregator operator name
+    this.sinkId = UUID.randomUUID().toString();
+  }
+
+  @Override
+  void init() {
+    super.init();
+    this.initTable = tableSupplier().get();
+    this.uidPrefix = builder().uidPrefix() == null ? "" : builder().uidPrefix();
+
+    // FlinkWriteConf properties needed to be set separately, so we do not have to serialize the
+    // full conf
+    this.upsertMode = flinkWriteConf().upsertMode();
+    this.dataFileFormat = flinkWriteConf().dataFileFormat();
+    this.targetDataFileSize = flinkWriteConf().targetDataFileSize();
+
+    this.commonCommitter =
+        new CommonCommitter(
+            tableLoader,
+            flinkWriteConf().branch(),
+            snapshotProperties,
+            flinkWriteConf().overwriteMode(),
+            flinkWriteConf().workerPoolSize(),
+            sinkId);
+  }
+
+  @Override
+  public DataStream<RowData> addPreWriteTopology(DataStream<RowData> inputDataStream) {
+    return distributeDataStream(inputDataStream);
+  }
+
+  @Override
+  public SinkV2Writer createWriter(InitContext context) {
+    RowDataTaskWriterFactory taskWriterFactory =
+        new RowDataTaskWriterFactory(
+            tableSupplier(),
+            flinkRowType(),
+            targetDataFileSize,
+            dataFileFormat,
+            writeProperties(),
+            equalityFieldIds(),
+            upsertMode);
+    IcebergStreamWriterMetrics metrics =
+        new IcebergStreamWriterMetrics(context.metricGroup(), initTable.name());
+    return new SinkV2Writer(
+        tableSupplier(),
+        taskWriterFactory,
+        metrics,
+        context.getSubtaskId(),
+        context.getAttemptNumber());
+  }
+
+  @Override
+  public DataStream<CommittableMessage<SinkV2Committable>> addPreCommitTopology(
+      DataStream<CommittableMessage<SinkV2Committable>> writeResults) {
+    TypeInformation<CommittableMessage<SinkV2Committable>> typeInformation =
+        CommittableMessageTypeInfo.of(this::getCommittableSerializer);
+    return writeResults
+        .global()
+        .transform(
+            prefixIfNotNull(uidPrefix, initTable.name() + "-" + sinkId + "-pre-commit-topology"),
+            typeInformation,
+            new SinkV2Aggregator(commonCommitter))
+        .uid(prefixIfNotNull(uidPrefix, "pre-commit-topology"))
```

Comment:
Maybe a better name/uid would be `...-write-result-aggregator` in case we add more things here later and it's also more descriptive